tests/native/test_state.py

```diff
@@ -211,6 +269,52 @@ def testContextSerialization(self):
         self.assertEqual(new_new_state.context["step"], 30)
 
 
+"""
+This function needs to be a global function for the following test or else we
+get the following error
+    E       AttributeError: Can't pickle local object 'StateHooks.test_state_hooks.<locals>.do_nothing'
+"""
+
+
+def do_nothing(_: StateBase) -> None:
+    return None
+
+
+def fin(_: StateBase) -> None:
+    print("Reached fin callback")
+    return None
+
+
+class StateHooks(unittest.TestCase):
+    def setUp(self):
+        core = config.get_group("core")
+        core.seed = 61
+        core.mprocessing = core.mprocessing.single
+
+        dirname = os.path.dirname(__file__)
+        self.m = Manticore(os.path.join(dirname, "binaries", "basic_linux_amd64"), policy="random")
+
+    def test_state_hooks(self):
+        @self.m.hook(0x400610, after=True)
+        def process_hook(state: State) -> None:
+            # We can't remove because the globally applied hooks are stored in
+            # the Manticore class, not State
+            self.assertFalse(state.remove_hook(0x400610, process_hook, after=True))
+            # We can remove this one because it was applied specifically to this
+            # State (or its parent)
+            self.assertTrue(state.remove_hook(None, do_nothing, after=True))
+
+            state.add_hook(None, do_nothing, after=False)
+            state.add_hook(None, do_nothing, after=True)
+            state.add_hook(0x400647, fin, after=True)
+            state.add_hook(0x400647, fin, after=False)
+
+        for state in self.m.ready_states:
+            self.m.add_hook(None, do_nothing, after=True, state=state)
+
+        self.m.run()
```

Comment:
If I'm reading it right, these tests cover whether the hooks can be registered correctly, but not whether they actually fire. Since one of the hooks here uses a `print` statement, perhaps we could wrap this line in a `contextlib.redirect_stdout` and check that "Reached fin callback" is present somewhere?

Reply:
Ahhh good catch. Could have sworn I intended to do that in the first place. Thank you 🙂